marker_error/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+}
+
+/// The enum of all errors. It has special variants with `Uncategorized` in their
+/// name which are used for general unrecoverable errors that we don't need to
+/// match on.
+///
+/// If some kind of error needs to be handled specially, it needs to be
+/// moved from `Uncategorized` to a dedicated enum variant in [`ErrorCategory::Categorized`].
+/// For example, if you want to add a help message or for some other elaborate cases.
+#[derive(Debug, thiserror::Error)]
+enum ErrorCategory<Kind> {
+    #[error(transparent)]
+    Categorized(Kind),
+
+    /// Used for errors that should just be propagated to the caller
+    /// with an additional `message` context.
+    #[error("{message}")]
+    Uncategorized {
+        message: String,
+        source: Option<Box<dyn std::error::Error + Send + Sync>>,
+    },
+
+    /// Used for errors that should just be propagated to the caller
+    /// without any additional context.
+    #[error(transparent)]
+    TransparentUncategorized(#[from] Box<dyn std::error::Error + Send + Sync>),
+
+    /// Used for errors that should just be propagated to the caller
+    /// and there are many related errors that should be propagated.
+    #[error("{message}")]
+    ManyUncategorized { message: String, errors: Vec<Error<Kind>> },
+}
+
+impl<Kind> Error<Kind> {
+    /// Output the error to `stderr`. Use this to ultimately display it to
+    /// the user in a top-level main function, for example.
+    pub fn print(&self)
+    where
+        Kind: miette::Diagnostic,
+    {
+        eprint!("{}", ErrorRender(self));
+    }
+
+    /// Get a reference to the error's typed kind. If this contains an uncategorized
+    /// error then [`None`] is returned.
+    pub fn kind(&self) -> Option<&Kind> {
+        match &self.imp.category {
+            ErrorCategory::Categorized(kind) => Some(kind),
+            _ => None,
+        }
+    }
+
+    /// Create a new categorized error from the given kind.
+    pub fn from_kind(kind: Kind) -> Self {
+        Self::from_category(ErrorCategory::Categorized(kind))
+    }
+
+    fn from_category(category: ErrorCategory<Kind>) -> Self {
+        // We don't use any closures to not add more stack frames to the backtrace
+        let trace = if std::env::var(MARKER_ERROR_TRACE).is_ok() {
+            Some(ErrorTrace::capture())
+        } else {
+            None
+        };
+
+        let imp = ErrorImp { category, trace };
+
+        Self { imp: Arc::new(imp) }
+    }
+
+    /// Wrap an existing error making it a source of this error, and add
+    /// more context with a message.
+    pub fn wrap(source: impl Into<Box<dyn std::error::Error + Send + Sync>>, message: impl Into<String>) -> Self {
+        Self::from_category(ErrorCategory::Uncategorized {
+            message: message.into(),
+            source: Some(source.into()),
+        })
+    }
+
+    /// Make a root error that doesn't have a source error
+    pub fn root(message: impl Into<String>) -> Self {
```

Comment:
Yes, there is a huge difference. `impl Into<String>`. The difference is that `into` consumes the value, but `to_string()` takes it by referece
```rust
fn into(self) -> String;
// vs
fn to_string(&self) -> String;
```
It means `impl Into<String>` doesn't do an allocation if you pass it an owned `String`, because such conversion would be noop. But if you pass `String` to `ToString`, it will always make a clone